File: app/src/main/java/com/yeraygarcia/recipes/database/dao/IngredientDao.java
package com.yeraygarcia.recipes.database.dao;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Query;

import com.yeraygarcia.recipes.database.entity.Ingredient;

import java.util.ArrayList;
import java.util.List;

@Dao
public abstract class IngredientDao extends BaseDao<Ingredient> {

    @Query("DELETE FROM ingredient")
    public abstract void deleteAll();

    @Query("SELECT * FROM ingredient")
    public abstract LiveData<List<Ingredient>> findAll();

    @Query("DELETE FROM ingredient WHERE id NOT IN (:ids)")
    abstract void deleteIfIdNotIn(List<Long> ids);

    public void deleteIfNotIn(List<Ingredient> entities) {
        List<Long> ids = new ArrayList<>();
        for (Ingredient entity : entities) {
            ids.add(entity.getId());
        }
        deleteIfIdNotIn(ids);
    }

    @Query("SELECT name FROM ingredient UNION SELECT name_singular FROM unit UNION SELECT name_plural FROM unit")
    public abstract LiveData<List<String>> getUnitsAndIngredientNames();

    @Query("SELECT name FROM ingredient")
    public abstract LiveData<List<String>> getIngredientNames();

    @Query("SELECT id FROM ingredient WHERE name = :name")
    public abstract Long getIngredientIdByName(String name);
}

File: app/src/main/java/com/yeraygarcia/recipes/database/dao/IngredientDao.kt
package com.yeraygarcia.recipes.database.dao

import android.arch.lifecycle.LiveData
import android.arch.persistence.room.Dao
import android.arch.persistence.room.Query
import com.yeraygarcia.recipes.database.entity.Ingredient
import java.util.*

@Dao
abstract class IngredientDao : BaseDao<Ingredient>() {

    @get:Query("SELECT name FROM ${IngredientDao.TABLE} UNION SELECT name_singular FROM ${UnitDao.TABLE} UNION SELECT name_plural FROM ${UnitDao.TABLE}")
    abstract val unitsAndIngredientNames: LiveData<List<String>>

    @get:Query("SELECT name FROM ${IngredientDao.TABLE}")
    abstract val ingredientNames: LiveData<List<String>>

    @Query("DELETE FROM $TABLE")
    abstract fun deleteAll()

    @Query("DELETE FROM $TABLE WHERE id IN (:ids)")
    abstract fun deleteByIds(ids: List<UUID>)

    @Query("SELECT * FROM $TABLE")
    abstract fun findAll(): LiveData<List<Ingredient>>

    @Query("SELECT * FROM $TABLE WHERE id = :id")
    abstract fun findById(id: UUID): Ingredient

    @Query("SELECT id FROM $TABLE WHERE name = :name")
    abstract fun findIdByName(name: String): UUID?

    @Query("SELECT * FROM $TABLE WHERE name = :name")
    abstract fun findByName(name: String): Ingredient?

    @Query("DELETE FROM $TABLE WHERE id NOT IN (:ids)")
    internal abstract fun deleteIfIdNotIn(ids: List<UUID>)

    fun deleteIfNotIn(entities: List<Ingredient>) {
        val ids = ArrayList<UUID>()
        for (entity in entities) {
            ids.add(entity.id)
        }
        deleteIfIdNotIn(ids)
    }

    @Query("SELECT * FROM $TABLE WHERE modified > :lastSync")
    abstract fun findModified(lastSync: Long): List<Ingredient>

    fun insertIfAbsent(name: String): Ingredient {
        return findByName(name) ?: Ingredient(name = name).apply { insert(this) }
    }

    companion object {
        const val TABLE = "ingredient"
    }
}
